torch/onnx/_internal/exporter/_compat.py

```diff
@@ -52,52 +57,51 @@ def _from_dynamic_axes_to_dynamic_shapes(
     for input_name, axes in dynamic_axes.items():
+        # NOTE: torch.export.Dim requires strict min and max constraints, and it
+        # dpends on the traced model to provide the correct min and max values.
+        # We set max to 99999 to avoid the constraints violation error with the default int64 max.
+        # https://github.com/pytorch/pytorch/blob/32f585d9346e316e554c8d9bf7548af9f62141fc/test/export/test_export.py#L687
         if input_name in output_names:
             # User specified an output name as a dynamic axis, so we skip it
             continue
-        # input_name can be either from input_names or from the model inputs
-        if input_name not in input_names_to_model_inputs:
-            raise ValueError(
-                f"dynamic axis: {input_name} is not found in the input names: {input_names}"
-            )
-        model_input_name = input_names_to_model_inputs[input_name]
         if isinstance(axes, dict):
-            dynamic_shapes_to_exported_program[model_input_name] = {
-                k: torch.export.Dim(v) for k, v in axes.items()
+            # Dim needs to pass str.isidentifier()
+            dynamic_shapes[input_name] = {
+                k: torch.export.Dim(re.sub(r"[^A-Za-z_]", "", v), max=99999)
+                for k, v in axes.items()
             }
         elif isinstance(axes, list):
-            dynamic_shapes_to_exported_program[model_input_name] = {
-                k: torch.export.Dim(f"{model_input_name}_dim_{k}") for k in axes
+            dynamic_shapes[input_name] = {
+                k: torch.export.Dim(f"{input_name}_dim_{k}", max=99999) for k in axes
             }
         else:
-            raise TypeError(
-                f"dynamic_axes value must be either a dict or a list, but got {type(axes)}"
+            raise ValueError(
+                "Unsupported dynamic_axes format. Please provide a dict or a list."
             )
-    # torch.export.export needs static dim to present in dynamic_shapes
-    # for all input tensors, so we need to add them with None
-    for input_name in sig.parameters:
-        if input_name not in dynamic_shapes_to_exported_program:
-            dynamic_shapes_to_exported_program[input_name] = None  # type: ignore[assignment]
 
-    return dynamic_shapes_to_exported_program
+    for input_name in input_names:
+        if input_name not in dynamic_shapes:
+            dynamic_shapes[input_name] = None
+
+    # Order the inputs according to the signature of the model
+    sig = _signature(model)
+    inputs = []
+    for idx, param_name in enumerate(sig.parameters):
+        if idx < len(args):
+            inputs.append(args[idx])
+        elif param_name in kwargs:
+            inputs.append(kwargs[param_name])
+
+    # We need tree structure to represent dynamic_shapes
+    _, tree_structure = _pytree.tree_flatten(inputs)
+    dynamic_shapes = _pytree.tree_unflatten(dynamic_shapes.values(), tree_structure)
```

Comment:
so this actually works. Nice!

Reply:
Would it be possible to take this out as a helper function and create unit tests specifically for the flatten-unflatten process? This way we have a better coverage of different scenarios.

Reply:
Done